### PR TITLE
[WIP] Split the plugin into various autoplugins.

### PR DESCRIPTION
**Pull request by nrinaudo**
This is a first step towards #30, but still very much a WIP.

I believe this works as expected, but:
- it's a massive change to the code's organisation that maintainers might not approve of
- I'm not good enough with SBT to be confident there are no unexpected side effects
- the documentation is not up-to-date

I'd appreciate feedback or confirmation that I'm going in the right direction, and will be happy to update the README accordingly once done.

Basic usage is as follows:
- `UnidocPlugin` (root project): generates unified scaladoc
- `GenJavadocPlugin` (child project), `JavaUnidocPlugin` (root project): generates unified javadoc
- `GenJavadocExtraPlugin` (child project): publish javadoc rather than scaladoc
- `GenJavadoc[Extra]Plugin` (child project), `ScalaJavaUnidocPlugin` (root project): generates unified scaladoc & javadoc

I realise the names are unpleasantly long - so far, I've tried to mirror the current convention, but have no issue with changing them to shorter values.

### CODE REVIEW

**Comment by 2m** [comment]
This is awesome! I will try to review this more thoroughly as soon as I get some free bandwidth.

Feel free to change the names so they become more intuitive. This will be a breaking change for the plugin users, and more easy names will surely be welcome.

Given my two minutes of thought have you considered such names:

* `GenJavadocPlugin` child project, generates javadoc from scala code
* `ScalaUnidocPlugin` root project, generates unified scaladoc
* `JavaUnidocPlugin` root project, generates unified javadoc

To get both javadoc and scaladoc unified, both plugins `ScalaUnidocPlugin` and `JavaUnidocPlugin` would have to be enabled. In that case `unidoc` would generate both.

For the publish javadoc feature, I think the plugin should have that explicit in its name, like `PublishJavadocPlugin`.

Also, the fact that everyting ends in `Plugin` seems to verbose. Especially since these plugin names are used already next to the same word, like `enablePlugins(ScalaUnidocPlugin)`. Can we just drop the `Plugin` ending from the plugin names? Has any plugin done this before? //cc @eed3si9n

**Comment by nrinaudo** [comment]
Thanks!

Regarding your suggestion that enabling both `ScalaUnidocPlugin` and `JavaUnidocPlugin` should replace `ScalaJavaUnidocPlugin` - I agree entirely, but the aggregate plugin does need to do something more. I'm not entirely sure what that's for - I've just been moving code around so far - but there's [this line](https://github.com/sbt/sbt-unidoc/pull/31/files#diff-7eb19e1a0faced91b16c8fb73775f238R19).

**Comment by nrinaudo** [comment]
Worked it out. I believe everything works as you suggested (except for removing the `Plugin` at the end of a plugin's name - it's still part of sbt's best practice and I'll wait for @eed3si9n to weigh in on this before changing it).

There's still a bit of cleanup I'd like to do in the code, but I'd rather wait until someone confirms I'm not going into the wrong direction entirely before that :)

**Comment by eed3si9n** [comment]
@2m wrote:

> Also, the fact that everyting ends in `Plugin` seems to verbose. Especially since these plugin names are used already next to the same word, like `enablePlugins(ScalaUnidocPlugin)`.
> Can we just drop the Plugin ending from the plugin names? Has any plugin done this before?

I would actually recommend suffixing all auto plugins with `Plugin` as a convention, even if it feels a bit verbose. I've taken the approach with `AssemblyPlugin`, `BuildInfoPlugin` etc, and [Plugins Best Practices](http://www.scala-sbt.org/0.13/docs/Plugins-Best-Practices.html) examples.
It's partly because in build.sbt plugins are imported by default along with all other kinds of names.
I think `ScalaUnidoc` should represent a pure Scala object that someone can reuse, and plugin should be called `ScalaUnidocPlugin`.

**Comment by nrinaudo** [comment]
@eed3si9n agreed. sbt-unidoc is one of these examples where there's a concrete reason for this, too: if you drop the `Plugin` suffix, you have both a plugin and a configuration called `ScalaUnidoc` brought in scope automatically.

File: src/main/scala/sbtunidoc/BaseUnidocPlugin.scala [hunk 1]
```diff
@@ -0,0 +1,50 @@
+package sbtunidoc
+
+import sbt.Keys._
+import sbt._
+
+object BaseUnidocPlugin extends AutoPlugin {
```
Comment (eed3si9n) [inline src/main/scala/sbtunidoc/BaseUnidocPlugin.scala]: This should override def requires `sbt.plugins.JvmPlugin`.

File: src/main/scala/sbtunidoc/BaseUnidocPlugin.scala [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+  object autoImport extends UnidocKeys
+  import autoImport._
+
+  override def projectSettings = Seq(
+    unidoc in Compile := Seq.empty,
+    unidoc in Test := Seq.empty
+  )
+
+  def commonSettings(sc: Configuration): Seq[sbt.Def.Setting[_]] = Seq(
```
Comment (eed3si9n) [inline src/main/scala/sbtunidoc/BaseUnidocPlugin.scala]: I normally use the name `commonSettings` specifically for common settings in `build.sbt`. I think it's better to call this baseUniDocSettings, or something along the line.

File: src/main/scala/sbtunidoc/GenJavadocPlugin.scala [hunk 3]
```diff
@@ -0,0 +1,17 @@
+package sbtunidoc
+
+import sbt._
+import sbt.Keys._
+
+object GenJavadocPlugin extends AutoPlugin {
```
Comment (eed3si9n) [inline src/main/scala/sbtunidoc/GenJavadocPlugin.scala]: This should override def requires sbt.plugins.JvmPlugin.

File: src/main/scala/sbtunidoc/BaseUnidocPlugin.scala (hunk 1 above)
Comment (nrinaudo) [inline src/main/scala/sbtunidoc/BaseUnidocPlugin.scala]: Does that let me get rid of `Defaults.configSettings` in, for example:

```scala
def scalaUnidocTask(c: Configuration, sc: Configuration): Seq[sbt.Def.Setting[_]] =
  inConfig(c)(Defaults.configSettings ++ baseScalaUnidocTasks(sc)) ++ Seq(
    unidoc in sc ++= Seq((doc in c).value)
  )
```

Comment (eed3si9n) [inline src/main/scala/sbtunidoc/BaseUnidocPlugin.scala]: No, but it makes sure that unidoc plugins come after the default settings.

**Comment by nrinaudo** [comment]
Requested changes were implemented. Depending on your preferences, I can:
- leave the code as-is, update the README and consider things done
- clean up the code somewhat first - it works, but I feel some comments explaining the interactions of the various plugins might be in order (ScalaUnidocPlugin depending on BaseUnidocPlugin for default `unidoc` values, for instance).

**Comment by eed3si9n** [comment]
It's up to you.

**Comment by nrinaudo** [comment]
Is it alright for me to update the documentation to modern versions of `sbt-site` and `sbt-ghpages`?

**Comment by eed3si9n** [comment]
Yea. Go for it.

**Comment by nrinaudo** [comment]
Documentation up to date, with one tiny caveat: the sbt-doctest version number is not the one that'll actually be used by the next release, if and when it comes.

I also pushed a fix for a (stupid) mistake I made - `build.sbt` was modified for my convenience (in order to call `publishLocal` more comfortably) but should never have been pushed. This is now fixed. Apologies for messing up the history.

**Pull request merged**